// postgresql-client/cjs/data-types/JsonType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayJsonType = exports.JsonType = void 0;
const definitions_1 = require("../definitions");
exports.JsonType = {
    name: 'json',
    oid: definitions_1.DataTypeOIDs.json,
    jsType: 'string',
    parseBinary(v, options) {
        const content = v.toString('utf8');
        const fetchAsString = options.fetchAsString &&
            options.fetchAsString.includes(definitions_1.DataTypeOIDs.jsonb);
        if (fetchAsString)
            return content;
        return content ? JSON.parse(content) : undefined;
    },
    encodeText(v) {
        if (typeof v === 'object' || typeof v === 'bigint')
            return JSON.stringify(v);
        if (typeof v === 'boolean')
            return v ? 'true' : 'false';
        return '' + v;
    },
    parseText(v, options) {
        const fetchAsString = options.fetchAsString &&
            options.fetchAsString.includes(definitions_1.DataTypeOIDs.jsonb);
        if (fetchAsString)
            return v;
        return v ? JSON.parse(v) : null;
    },
    isType(v) {
        return v && typeof v === 'object';
    }
};
exports.ArrayJsonType = {
    ...exports.JsonType,
    name: '_json',
    oid: definitions_1.DataTypeOIDs._json,
    elementsOID: definitions_1.DataTypeOIDs.json
};
